### PR TITLE
[DAML on SQL] Validate party allocation [KVL-693]

### PR DESCRIPTION
**Pull request by fabiotudone-da**
CHANGELOG_BEGIN
- [DAML on SQL] Disallow --implicit-party-allocation
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by SamirTalwar-DA — COMMENTED** [review]
Hmmm. Instead of disallowing it, can't we move it out of `CommonCli` and into the two Sandbox `Cli` classes?

If they end up with too much duplication, we can then refactor that out.

**Comment by fabiotudone-da** [comment]
> Hmmm. Instead of disallowing it, can't we move it out of `CommonCli` and into the two Sandbox `Cli` classes?
> 
> If they end up with too much duplication, we can then refactor that out.

I agree it's valuable to do so but I was thinking about having a separate PR for that. @gerolf-da WDYT?

File: ledger/daml-on-sql/src/test/suite/scala/on/sql/CliSpec.scala
Comment (gerolf-da) [inline ledger/daml-on-sql/src/test/suite/scala/on/sql/CliSpec.scala:47]: The parameter `--implicit-party-allocation` is not available for the DAML on SQL CLI. This parse probably returns `None` because it cannot find a definition of that parameter.

**Comment by gerolf-da** [comment]
> > Hmmm. Instead of disallowing it, can't we move it out of `CommonCli` and into the two Sandbox `Cli` classes?
> > If they end up with too much duplication, we can then refactor that out.
> 
> I agree it's valuable to do so but I was thinking about having a separate PR for that. @gerolf-da WDYT?

It doesn't need to be moved out of `CommonCli`, because it's not in there to begin with, or am I missing something?

File: ledger/daml-on-sql/src/test/suite/scala/on/sql/CliSpec.scala
Comment (fabiotudone-da) [reply ledger/daml-on-sql/src/test/suite/scala/on/sql/CliSpec.scala]: Oh true, it is not parsed by CommonCli but by Sandbox next. Removed checks.

**Comment by fabiotudone-da** [comment]
> > > Hmmm. Instead of disallowing it, can't we move it out of `CommonCli` and into the two Sandbox `Cli` classes?
> > > If they end up with too much duplication, we can then refactor that out.
> > 
> > 
> > I agree it's valuable to do so but I was thinking about having a separate PR for that. @gerolf-da WDYT?
> 
> It doesn't need to be moved out of `CommonCli`, because it's not in there to begin with, or am I missing something?

@SamirTalwar were you referring to refactoring the config? I didn't think it out too deeply but I assumed you'd like to remove it completely from all ledgers except sandbox.

**Comment by SamirTalwar-DA** [comment]
> It doesn't need to be moved out of `CommonCli`, because it's not in there to begin with, or am I missing something?

Haha, I saw the changes and assumed they were doing something helpful. It didn't occur to me that that code path might never get hit, so I thought it must be in `CommonCli`. Should have checked.

Nice catch, @gerolf-da!

**Review by SamirTalwar-DA — COMMENTED** [review]
With this change, we should be able to remove `--open-world` and `--exclude=ClosedWorldIT` from the conformance tests. Right now there's no proof that this actually accomplishes anything.

**Comment by SamirTalwar-DA** [comment]
Yup, as I suspected. It goes far deeper than just a configuration option.

You will at the very least need to change `SandboxServer`, which overrides this value and hard-codes it to `true`. However, from what I recall when adding this code, this is set to `true` because setting it to `false` breaks stuff. There's probably more work to untangle this.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:138]: I don't like this `Await` here. Can we propagate the future outwards?

File: ledger/sandbox-classic/src/main/scala/platform/sandbox/SandboxServer.scala
Comment (SamirTalwar-DA) [inline ledger/sandbox-classic/src/main/scala/platform/sandbox/SandboxServer.scala:282]: Please name this parameter.

In fact, please name all the parameters. :upside_down_face:

File: ledger/sandbox-classic/src/main/scala/platform/sandbox/stores/SandboxIndexAndWriteService.scala
Comment (SamirTalwar-DA) [inline ledger/sandbox-classic/src/main/scala/platform/sandbox/stores/SandboxIndexAndWriteService.scala:59]: Does this need a default?

File: ledger/sandbox-classic/src/main/scala/platform/sandbox/stores/ledger/sql/SqlLedger.scala
Comment (SamirTalwar-DA) [inline ledger/sandbox-classic/src/main/scala/platform/sandbox/stores/ledger/sql/SqlLedger.scala:79]: Likewise, does this need a default?

If it's just for tests, I'd rather the tests were updated to be explicit.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:922]: Oh, here too. Do we need the default-ness?

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:138]: We probably can but I think it might not be easy and probably it's not worth it:

1. this change is on the previous DAML on SQL that is now only being maintained so I think it should quick, small and have the smallest impact possible
2. `PostCommitValidation.validate` was designed as sync but we can only reuse a query behind an async interface (and it doesn't make a lot of sense to create a new one)
3. `Await.result` will block the calling thread, in line with the sync/blocking nature of `PostCommitValidation.validate` so I think it's not so bad after all

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:922]: In order to minimize the impact (and, specifically, to restrict it to DAML-on-SQL as much as possible) I'd prefer to have a default at least in (non-test) factories. Why would you prefer to make it always explicit?

File: ledger/sandbox-classic/src/main/scala/platform/sandbox/stores/ledger/sql/SqlLedger.scala
Comment (fabiotudone-da) [reply ledger/sandbox-classic/src/main/scala/platform/sandbox/stores/ledger/sql/SqlLedger.scala:79]: It's not just for tests but I have made the value explicit in the test functions.

File: ledger/sandbox-classic/src/main/scala/platform/sandbox/SandboxServer.scala
Comment (fabiotudone-da) [reply ledger/sandbox-classic/src/main/scala/platform/sandbox/SandboxServer.scala]: Done.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:922]: If we find new uses for this code, the default can be misleading; it may not be appropriate for the new use case. I'd rather any configuration was made explicit at the call site, not the implementation site.

It's a minor thing though, so happy to defer to you on the final decision, as you have more familiarity with this code.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:138]: The caller is itself inside a `Future`, so it's blocking the execution context. This can be quite dangerous.

`PostCommitValidation#validate` is only called in one place (except for tests), and could easily be transformed to return a `Future[Option[RejectionReason]]`.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:922]: I agree with your point in general but classic (non-KV) SQL-backed drivers should not perform party allocation validation, else e.g. SDK tools like the sandbox might break. The exceptions are very intentional and specific and, currently, only DAML on SQL.

I'll try having a look at the impacts of making it explicit though, maybe they are minor.

Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:922]: Specific always trumps general. :slightly_smiling_face:

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:138]: This is pre-existing though and has not caused problems AFAIK but I'll refactor it, only I'd do that in a separate PR (but right after merging this, so we don't forget). Would that be OK?

Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:138]: Sold!

Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:138]: BTW, consider that it hasn't caused problems because up until now, there have been no `Await` calls here. In other words, it has been very fast.

**Review by SamirTalwar-DA — APPROVED** [review]

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:922]: It looks like it would need to be propagated at least in platform/index and platform/store, i.e. `participant-integration-api`, not even sure we want/are allowed to do that. WDYT?

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:138]: It seems to have been compute-bound so far, yes. OK, that being the case then I actually think it's better to tackle this as part of this PR.

Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:138]: Nah, this too has far-reaching effects, up to the Runner. It'd better be done here but that would inflate this PR too much.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:922]: Anyway, I think this refactoring too is better done in a separate PR, if at all.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala
Comment (fabiotudone-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:138]: Created #7972.

**Pull request merged**
